server/protocols/webadmin/src/test/java/org/apache/james/webadmin/routes/UsersRoutesTest.java

```diff
@@ -0,0 +1,434 @@
+        }
+
+        @Test
+        public void postShouldReturnUserErrorWhenNoBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(400);
+        }
+
+        @Test
+        public void postShouldReturnUserErrorWhenEmptyJsonBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(400);
+        }
+
+        @Test
+        public void postShouldReturnUserErrorWhenWrongJsonBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{\"bad\":\"any\"}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(400);
+        }
+
+        @Test
+        public void postShouldReturnUserErrorWhenMissingPasswordJsonBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{\"username\":\"username\"}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(400);
+        }
+
+        @Test
+        public void postShouldReturnUserErrorWhenMissingUserJsonBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{\"password\":\"password\"}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(400);
+        }
+
+        @Test
+        public void postShouldReturnOkWhenValidJsonBody() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{\"username\":\"username@domain\",\"password\":\"password\"}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(200);
+        }
+
+        @Test
+        public void postShouldAddTheUser() {
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+                .body("{\"username\":\"" + USERNAME + "\",\"password\":\"password\"}")
+            .when()
+                .post(UserRoutes.USERS)
+            .then()
+                .statusCode(200);
+
+            given()
+                .accept(ContentType.JSON)
+                .contentType(ContentType.JSON)
+            .when()
+                .get(UserRoutes.USERS)
+            .then()
+                .statusCode(200)
+                .body(equalTo("[\"" + USERNAME + "\"]"));
+        }
+
+        @Test
+        public void postingTwoTimesShouldBeAllowed() {
```

Comment:
Covered by the 500 error.

Scenario : To be concurrent, both requests have to believe there is not entry (the user check returns null)

Then one request will create the user before the other one. The one created will get a 200 code, the other one, not applied will get a 500.

If not in this scenario (which is acceptable), even with linearizability you will have a conflict as operation do not commute. Thus, I will not detail this case ;-)